core/src/main/java/org/apache/iceberg/io/inmemory/InMemoryCatalog.java

```diff
@@ -0,0 +1,374 @@
+    return SLASH.join(defaultNamespaceLocation(tableIdentifier.namespace()), tableIdentifier.name());
+  }
+
+  private String defaultNamespaceLocation(Namespace namespace) {
+    if (namespace.isEmpty()) {
+      return warehouseLocation;
+    } else {
+      return SLASH.join(warehouseLocation, SLASH.join(namespace.levels()));
+    }
+  }
+
+  @Override
+  public Table createTable(
+      TableIdentifier tableIdentifier,
+      Schema schema,
+      PartitionSpec spec,
+      String location,
+      Map<String, String> properties) {
+
+    // Create namespace automatically if it does not exist.
+    if (!namespaceExists(tableIdentifier.namespace())) {
+      createNamespace(tableIdentifier.namespace());
+    }
+
+    return super.createTable(tableIdentifier, schema, spec, location, properties);
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier tableIdentifier, boolean purge) {
+    TableOperations ops = newTableOps(tableIdentifier);
+    TableMetadata lastMetadata;
+    if (purge && ops.current() != null) {
+      lastMetadata = ops.current();
+    } else {
+      lastMetadata = null;
+    }
+
+    if (tableDb.remove(tableIdentifier) == null) {
+      LOG.info("Table does not exist: cannot drop table {}", tableIdentifier);
+      return false;
+    }
+
+    if (purge && lastMetadata != null) {
+      CatalogUtil.dropTableData(ops.io(), lastMetadata);
+    }
+
+    return true;
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    if (!namespaceExists(namespace) && !namespace.isEmpty()) {
+      throw new NoSuchNamespaceException(
+          "Namespace does not exist: cannot list tables for namespace %s", namespace);
+    }
+
+    return tableDb.keySet().stream()
+        .filter(t -> namespace.isEmpty() || t.namespace().equals(namespace))
+        .sorted(Comparator.comparing(TableIdentifier::toString))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public void renameTable(TableIdentifier fromTableIdentifier, TableIdentifier toTableIdentifier) {
+    if (fromTableIdentifier.equals(toTableIdentifier)) {
+      return;
+    }
+
+    if (!namespaceExists(toTableIdentifier.namespace())) {
+      throw new NoSuchNamespaceException("Cannot rename %s to %s because the namespace %s does not exist",
+          fromTableIdentifier, toTableIdentifier, toTableIdentifier.namespace());
+    }
+
+    if (!tableDb.containsKey(fromTableIdentifier)) {
+      throw new NoSuchTableException("Cannot rename %s to %s because the table %s does not exist",
+          fromTableIdentifier, toTableIdentifier, fromTableIdentifier);
+    }
+
+    if (tableDb.containsKey(toTableIdentifier)) {
+      throw new AlreadyExistsException("Cannot rename %s to %s because the table %s already exists",
+          fromTableIdentifier, toTableIdentifier, toTableIdentifier);
+    }
+
+    String fromLocation = tableDb.remove(fromTableIdentifier);
```

Comment:
If we intend this to support concurrent access, we should be synchronizing on the `tabldDB` for operations like this.